### PR TITLE
added elitism - no mutations for the best human

### PR DESCRIPTION
**Pull request by leshchenko1979**
Hi!

I noticed that your algorithm lacked so called "elitism" - a technique whereby the fittest network gets carried over to the next generation without mutations. It usually makes the network learn faster.

This pull request a few lines of code to implement that.

It also fixes what seemed to be a bug in the selectOne() function - now parents are more likelier to be chosen for crossover the fitter they are.

### CODE REVIEW

File: src/ga.js [hunk 1]
```diff
@@ -122,15 +126,13 @@ class GeneticAlgorithm {
...
-
-    index -= 1;
-    return globals.humans[index];
+    do {
+      let r = Math.random();
+      var betterHumans = globals.humans.filter(h => h.fitness > r);
+    } while (betterHumans.length == 0);
+    
+    let one = betterHumans[Math.floor(Math.random()*betterHumans.length)];
+    return one;
```
Comment (adityathebe) [inline src/ga.js:135]: I think this algorithm is slower than the existing one due to array manipulations.

̶A̶n̶y̶ ̶r̶e̶a̶s̶o̶n̶ ̶w̶h̶y̶ ̶y̶o̶u̶ ̶c̶h̶a̶n̶g̶e̶d̶ ̶i̶t̶ ̶?̶

**EDIT**: I'll look into it

File: src/ga.js [hunk 2]
```diff
@@ -84,7 +84,11 @@ class GeneticAlgorithm {
 
     // Create New Set of humans
     const newGeneration = [];
-    for (let i = 0; i < config.populationSize; i++) {
+    
+    //Elitism: add the clone of the best human to the new generation without mutations for faster neural network learning
```
Comment (adityathebe) [inline src/ga.js:88]: Thank you. I didn't know about Elitism.

Did you run some simulations with this added ? Any progress ?

Comment (leshchenko1979) [reply src/ga.js]: I did. It surely helps the evolution not to get bogged in random mutations and move steadily towards the goal.

However, the introduction of motor noise makes even the best human from the previous run perform poorly on the next one and the evolution loses all the valuable mutations as he gets eliminated.

Have you tried reducing the motor noise?

Comment (adityathebe) [reply src/ga.js:88]: > Have you tried reducing the motor noise?

I haven't. Will give it a try

File: src/ga.js (hunk 1 above)
Comment (adityathebe) [reply src/ga.js:135]: @leshchenko1979 I did some tests and you were right. The selectOne() function wasn't performing very well.

Thanks for contributing :)

**Pull request merged**